hack/deployer/runner/k3d.go

```diff
@@ -0,0 +1,164 @@
+
+package runner
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/v3/hack/deployer/exec"
+	"github.com/elastic/cloud-on-k8s/v3/hack/deployer/runner/env"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/vault"
+)
+
+const (
+	K3dDriverID = "k3d"
+
+	DefaultK3dRunConfigTemplate = `id: k3d-dev
+overrides:
+  clusterName: %s-dev-cluster
+`
+)
+
+func init() {
+	drivers[K3dDriverID] = &K3dDriverFactory{}
+}
+
+type K3dDriverFactory struct{}
+
+var _ DriverFactory = &K3dDriverFactory{}
+
+func (k K3dDriverFactory) Create(plan Plan) (Driver, error) {
+	return &K3dDriver{
+		plan:        plan,
+		vaultClient: vault.NewClientProvider(),
+		clientImage: plan.K3d.ClientImage,
+		nodeImage:   plan.K3d.NodeImage,
+	}, nil
+}
+
+type K3dDriver struct {
+	plan        Plan
+	clientImage string
+	vaultClient vault.ClientProvider
+	nodeImage   string
+}
+
+func (k *K3dDriver) Execute() error {
+	switch k.plan.Operation {
+	case CreateAction:
+		return k.create()
+	case DeleteAction:
+		return k.delete()
+	}
+	return nil
+}
+
+func (k *K3dDriver) create() error {
+	cmd := k.cmd("cluster", "create", "--image", k.plan.K3d.NodeImage)
+	if cmd == nil {
+		return fmt.Errorf("failed to create k3d cluster")
+	}
+	err := cmd.Run()
+	if err != nil {
+		return err
+	}
+
+	// Get kubeconfig from k3d
+	kubeCfg, err := k.getKubeConfig()
+	if err != nil {
+		return err
+	}
+	defer os.Remove(kubeCfg.Name())
+
+	if err := kubectl("--kubeconfig", kubeCfg.Name(), "delete", "storageclass", "standard"); err != nil {
+		return err
+	}
+
+	tmpStorageClass, err := k.createTmpStorageClass()
+	if err != nil {
+		return err
+	}
+
+	return kubectl("--kubeconfig", kubeCfg.Name(), "apply", "-f", tmpStorageClass)
+}
+
+func (k *K3dDriver) delete() error {
+	return fmt.Errorf("unimplemented")
+}
+
+func (k *K3dDriver) cmd(args ...string) *exec.Command {
+	params := map[string]interface{}{
+		"ClusterName":    k.plan.ClusterName,
+		"SharedVolume":   env.SharedVolumeName(),
+		"K3dClientImage": k.clientImage,
+		"K3dNodeImage":   k.nodeImage,
+		"Args":           args,
+	}
+
+	// on macOS, the docker socket is located in $HOME
+	dockerSocket := "/var/run/docker.sock"
+	if runtime.GOOS == "darwin" {
+		dockerSocket = "$HOME/.docker/run/docker.sock"
+	}
+	// We need the docker socket so that kind can bootstrap
+	// --userns=host to support Docker daemon host configured to run containers only in user namespaces
+	command := `docker run --rm \
+		--userns=host \
+		-v {{.SharedVolume}}:/home \
+		-v /var/run/docker.sock:` + dockerSocket + ` \
+		-e HOME=/home \
+		-e PATH=/ \
+		{{.K3dClientImage}} \
+		{{Join .Args " "}} {{.ClusterName}}`
+	cmd := exec.NewCommand(command)
+	cmd = cmd.AsTemplate(params)
+	return cmd
+}
+
+func (k *K3dDriver) getKubeConfig() (*os.File, error) {
+	// Get kubeconfig from k3d binary
+	output, err := k.cmd("kubeconfig", "get").WithoutStreaming().Output()
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
I think it would be good if we could run this locally too. Because we are bootstrapping k3d from inside a container it uses host.docker.internal as the API_HOST. I think this is fine in CI because we are also running the whole e2e command from inside a docker container. 

I could not think for something better than this very crude replacement: 

```
	// Replace host.docker.internal with 127.0.0.1 only if not running inside the CI container
	if os.Getenv("CI") != "true" {
		output = strings.Replace(output, "host.docker.internal", "127.0.0.1", -1)
	}
```

But maybe you have better idea?

Reply:
I tested this on both macos, and linux and this only was needed on macos, so I also am checking the `runtime.GOOS` here.